overview of examples / topology optimization example (#531)
Ferrite-FEM/Ferrite.jl: docs/src/overview.md

 A program for the use of linear shell elements is set up. The theoretical background is introduced as well.  
	
 Keywords: Shell elements, displacements, rotations
 
11.	[Nearly Incompressible Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/quasi_incompressible_hyperelasticity/)  

 This program combines ideas from the examples Incompressible Elasticity and Hyperelasticity to construct a mixed element solving three-dimensional displacement-pressure equations.
	
 Keywords: Non-linear problem, hyperelasticity, finite strain, large deformations, Newton's method, automatic differentiation, coupled problem, mixed elements, displacement-pressure equation, incompressibility

12.	[Incompressible Navier-Stokes Equations via DifferentialEquations.jl](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/ns_vs_diffeq/)  

 The weak form of the semi-discretized incompressible Navier-Stokes equations are derived from the strong form and implemented in the finite element code. Then, the time-dependent solution is calculated by using the DifferentialEquations.jl package. For this purpose, the PDE is required in a specific form.  
	

>  Keywords: Fluid dynamics, weak and strong form, solver, incompressibility, time-dependent problem

Reviewer: Can we move this into a separate list below the "pure" Ferrite examples? This is technically an example of how to integrate Ferrite.jl with DifferentialEquations.jl and qualitatively different from the other examples.

Author: Yes!

Author: changed in current version

```@meta
DocTestSetup = :(using Ferrite)
```

# Documentation of examples

On this page, you find a documentation for each example currently available for `Ferrite`. The collection contains programs displaying different aspects and techniques of the `Ferrite` toolbox. 
The programs are sorted as

* A complete list shortly summarizing the topic, what they teach and which keywords related to the task they contain
* Grouped by topic 

## List of examples (chronological order)
1.	[Heat Equation](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/heat_equation/)  

>  This example is the easiest way for getting started. The heat equation is solved on a unit square. For this purpose, a program is set up containing all the important aspects of finite element computation, including a short introduction to weak and strong forms, trial and test functions, degrees of freedom, boundary conditions, element assembly, global assembly, solving the system and visualizing the results.  

Reviewer 2: ```suggestion
 This example presents the easiest way to get started. The heat equation is solved on a unit square. For this purpose, a program is set up containing the core aspects of finite element computation: weak and strong forms, trial and test functions, degrees of freedom, boundary conditions, element routine, global assembly, solving the system and visualizing the results.
```

 
 Keywords: Fundamentals, heat equation, weak and strong form, Dirichlet boundary condition, assembly

2.	[Postprocessing](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/postprocessing/)  

 Based on example 1, visualization of the flux variable is done via the L2Projector. Additionally, the point evaluation along a line is shown.  
	

>  Keywords: Fundamentals, postprocessing, heat equation, flux, L2projection, point Evaluation

Reviewer 2: ```suggestion
 Keywords: Fundamentals, postprocessing, heat equation, flux, L2-projection, point evaluation
```

3.	[Helmholtz equation](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/helmholtz/)  

 The Helmholtz equation is solved on a unit square. Dirichlet and Neumann boundary conditions are applied at different parts of the boundary. A known analytical solution is approximately reproduced to verify the finite element solution.  
	

>  Keywords: Fundamentals, Helmholtz equation, Dirichlet boundary condition, Neumann boundary condition, verification

Reviewer 2: ```suggestion
 Keywords: Fundamentals, Helmholtz equation, Dirichlet boundary condition, Neumann boundary condition, verification, boundary integrals
```

4.	[Incompressible Elasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/incompressible_elasticity/)  

 A mixed element for solving unidirectional displacement-pressure coupling is constructed. The solution for nearly incompressible materials is compared for different interpolations.  
	

>  Keywords: Mixed elements, unidirectional coupling, displacement-pressure equation, incompressibility, stability

Author: ```suggestion
 Keywords: Mixed elements, multiple fields, unidirectional coupling, displacement-pressure equation, incompressibility, stability
```

>  Keywords: Mixed elements, unidirectional coupling, displacement-pressure equation, incompressibility, stability

Reviewer 2: ```suggestion
 Keywords: Mixed elements, unidirectional coupling, displacement-pressure equation, incompressibility, stability, exploiting system symmetry
```

Author: I don't think it is necessary to introduce it as a keyword considering the trick is done in many examples and only improves performance very slightly

Reviewer 2: I was thinking of this example as a fundamental one introducing this aspect.
Thinking of it in another category, I would agree that this key word is not needed.

5.	[Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/hyperelasticity/)  

>  A hyperelastic material model is solved in a finite strain setting. Stress and material tangent are recovered by using automatic differentiation. Newton’s method is used to iteratively solve the resulting non-linear system.

Reviewer 2: ```suggestion
 A hyperelastic material model is solved in a finite strain setting. Stress and material tangents are recovered by using automatic differentiation. Newton’s method is used to iteratively solve the resulting non-linear system.
```

Author: In this example I would consider the resulting stiffness matrix as a single material tangent, as it is not a coupled problem.

Reviewer 2: Makes sense, just leave it as it is.
(I am just used to the plural form, that's it.)

	
 Keywords: Plasticity, 3D, material modeling, material state, non-linear problem, Newton’s method, 

8.	[Time Dependent Problems](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/transient_heat_equation/)
  
 The transient heat equation is solved on a rectangular plate. The time discretization is done by using the implicit Euler scheme.  
	
 Keywords: Time dependent problem, transient heat equation, implicit Euler scheme

9.	[Ginzburg-Landau model energy minimization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/landau/)  
  
 A basic Ginzburg-Landau model is solved by combining `Ferrite` with `ForwardDiff`. Using threads, the calculation time is optimized.  
	
 Keywords: Gizburg-Landau, ForwardDiff.jl, parallelization, optimization, performance

10.	[Linear shell](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/linear_shell/)  

 A program for the use of linear shell elements is set up. The theoretical background is introduced as well.  
	

>  Keywords: Shell elements, displacements, rotations

Reviewer 2: ```suggestion
 Keywords: Shell elements, displacements, rotations, ForwardDiff.jl, under integration
```

>  Keywords: Shell elements, displacements, rotations, ForwardDiff.jl, under integration

Reviewer: Just a quick question - isn't the correct translation here "reduced integration" or am I simply misunderstanding the wording?

Reviewer 2: I copied that term from the example.
Personally, I would call it "selectively reduced integration" or "reduced integration"

Author: I think you are correct if I am not mistaken, however the same wording is used in the documentation of the example.

Author: ```suggestion
 Keywords: Shell elements, displacements, rotations, multiple fields, ForwardDiff.jl, reduced integration
```

 
11.	[Nearly Incompressible Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/quasi_incompressible_hyperelasticity/)  

 This program combines ideas from the examples Incompressible Elasticity and Hyperelasticity to construct a mixed element solving three-dimensional displacement-pressure equations.
	

>  Keywords: Non-linear problem, hyperelasticity, finite strain, large deformations, Newton's method, automatic differentiation, coupled problem, mixed elements, displacement-pressure equation, incompressibility

Author: ```suggestion
 Keywords: Non-linear problem, hyperelasticity, finite strain, large deformations, Newton's method, automatic differentiation, coupled problem, mixed elements, multiple fields, displacement-pressure equation, incompressibility
```

12.	[Computational homogenization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/computational_homogenization/)  

 A two-dimensional representative volume element (RVE) is loaded in shear. Dirichlet and periodic boundary conditions are applied and the results from the homogenization are compared to the Voigt and Reuss bounds.  
	
 Keywords: homogenization, periodic boundary conditions, representative volume element (RVE), microscale, Voigt and Reuss bound

13.	[Stokes flow](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/stokes-flow/)  

 Stokes flow on a quarter circle is solved by applying periodic boundary conditions. The weak and strong form of the problem are discussed including constrains. Mesh generation is done directly via the Gmsh API.  
	

>  Keywords: Periodic boundary conditions, multiple fields, mean value constraint, mesh generation with Gmsh, coupled problem, weak and strong form

Reviewer 2: ```suggestion
 Keywords: Periodic boundary conditions, mean value constraint, mesh generation with Gmsh, coupled problem, weak and strong form
```

Author: I would like to keep the keyword "multiple fields" to show that the DofHandler here consists of different parts.

Reviewer 2: And should we add the same key word for other examples with multiple fields?

Author: Yes, done!

 
14.	[Topology Optimization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/topology_optimization/)  

 Topology optimization is shown for the bending problem by using a SIMP material model. To avoid numerical instabilities, a regularization scheme requiring the calculation of the Laplacian is imposed, which is done by using the grid topology functionalities.  
	
 Keywords: Topology optimization, weak and strong form, non-linear problem, Laplacian, grid topology

## Other examples
1.	[Incompressible Navier-Stokes Equations via DifferentialEquations.jl](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/ns_vs_diffeq/)  

 The weak form of the semi-discretized incompressible Navier-Stokes equations are derived from the strong form and implemented in the finite element code. Then, the time-dependent solution is calculated by using the DifferentialEquations.jl package. For this purpose, the PDE is required in a specific form.  
	
 Keywords: Fluid dynamics, weak and strong form, solver, incompressibility, time-dependent problem

## Grouped by topic

> * Fundamentals: [Heat Equation](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/heat_equation/), [Postprocessing](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/postprocessing/), [Helmholtz Equation](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/helmholtz/)

Reviewer 2: ```suggestion
* Fundamentals: [Heat Equation](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/heat_equation/), [Postprocessing](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/postprocessing/), [Helmholtz Equation](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/helmholtz/), [Incompressible Elasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/incompressible_elasticity/), [Computational homogenization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/computational_homogenization/)
```

Reviewer 2: I just thought that all examples should be listed in some category.
That's why I added the ones that were not listed in any category yet.
I think that it would be nice to have an examples under "Fundamentals" which shows how to treat multiple fields.
(Even if it is pretty straight forward.) And "Incompressible Elasticity" would be a good choice for that, I think.

Regarding "Computational Homogenization" you are probably right: Its too specific for "Fundamentals".

Author: I agree about the part that every example should be listed at least in one category. If not category is a good fit, we should add a category "other examples".

Personally, I would consider using FEM with multiple fields and coupling to be already too advanced for "Fundamentals". That does not mean it has to be a very hard problem, but most students / beginners should start with simpler problems in my opinion, since there are already a lot of additional steps to consider for coupled problems.

Reviewer 2: Maybe we could have a category "Coupled Problems" and list this example as the first one?

Author: I think I will add two categories: "Multiple fields" to account for all problems which are either fully coupled, unidirectional coupled or contain multiple fields for the calculation (shells) and "Periodic problems" for the examples Comp. homogenization and stokes flow. This way, each example is present at least once in the Grouped by topic section.

Reviewer: I think we should be a bit careful regarding the wording. I would classify "coupled problems" as a subsection of problems containing multiple fields, where, loosely speaking, different sets of fields have different underlying physics (e.g. fluid-stucture interacton). Incompressibility on the other hand acts as a constraint on your PDE. Certainly, incompressibility is advanced from a purely theoretical perspective, but apart from this is the algorithmic treatment of incompressible linear elasticity rather straight forward. 

However, I agree hat we might want to have a simpler first example on multiple fields that does not have this large theoretical building (for users without background in mechanics). I can set up some reaction-diffusion system as a simple first example for multiple fields here for the future, if this is what we need.

Author: ```suggestion
* Fundamentals: [Heat Equation](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/heat_equation/), [Postprocessing](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/postprocessing/), [Helmholtz Equation](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/helmholtz/), [Incompressible Elasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/incompressible_elasticity/)
```

* Non-linear Problems: [Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/hyperelasticity/), [von Mises Plasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/plasticity/), [Nearly Incompressible Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/quasi_incompressible_hyperelasticity/), [Topology Optimization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/topology_optimization/)
* Time dependent problems: [Time Dependent Problems](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/transient_heat_equation/), [Incompressible Navier-Stokes Equations via DifferentialEquations.jl](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/ns_vs_diffeq/)

> * Advanced: [Threaded Assembly](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/threaded_assembly/), [Ginzburg-Landau model energy minimization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/landau/)

Reviewer 2: ```suggestion
* Advanced: [Threaded Assembly](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/threaded_assembly/), [Ginzburg-Landau model energy minimization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/landau/), [Linear shell](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/linear_shell/), [Stokes flow](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/stokes-flow/)
```

Author: In my opinion, the Threaded Assembly example is the only truely "advanced" example in the sense that it only improves the speed of the program, making it optional for someone who does not care about performance too much. In contrast, all other examples showcase something new and different.

My stance on this is of course subjective and I would ask for addtional opinions.

Reviewer: I think that, from a technical point of view, the Ginzburg-Landau model is an advanced modeling example that combines some of the techniques of other examples to speed up the computation. However, the threaded assembly is an example showing advanced algorithms. Not sure if the distinction is worth to make at the moment, because we just have a few examples for now. Stokes is the basic example for periodic boundary condtitions and mean value constraint, while the linear shell tells you how to deal with shells. Would not consider these two advanced either.

> * Time dependent problems: [Time Dependent Problems](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/transient_heat_equation/), [Incompressible Navier-Stokes Equations via DifferentialEquations.jl](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/ns_vs_diffeq/)
> * Advanced: [Threaded Assembly](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/threaded_assembly/), [Ginzburg-Landau model energy minimization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/landau/)

Author: ```suggestion
* Time dependent problems: [Time Dependent Problems](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/transient_heat_equation/), [Incompressible Navier-Stokes Equations via DifferentialEquations.jl](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/ns_vs_diffeq/)
* Multiple fields: [Incompressible Elasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/incompressible_elasticity/), [Linear shell](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/linear_shell/), [Nearly Incompressible Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/quasi_incompressible_hyperelasticity/), [Stokes flow](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/stokes-flow/)    
* Periodic problems: [Computational homogenization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/computational_homogenization/), [Stokes flow](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/stokes-flow/)
* Advanced: [Threaded Assembly](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/threaded_assembly/), [Ginzburg-Landau model energy minimization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/landau/)
```

Author: (reasoning see above)

```@meta
DocTestSetup = :(using Ferrite)
```

# Documentation of examples

On this page, you find a documentation for each example currently available for `Ferrite`. The collection contains programs displaying different aspects and techniques of the `Ferrite` toolbox. 
The programs are sorted as

* A complete list shortly summarizing the topic, what they teach and which keywords related to the task they contain
* Grouped by topic 

> ## List of examples (chronological order)

Reviewer 3: ```suggestion
## List of examples
```
I think the list order looks good, but e.g. postprocessing was added after the von Mises example.
Super nice work @blaszm - I need to study that topology example closer - the method does look very neat!
(Maybe it could make sense to move "Nearly Incompressible Hyperelasticity" after "Hyperelasticity" since it follows from "Incompressible elasticity" and "Hyperelasticity"?)

Reviewer: Ah right, the  "Nearly Incompressible Hyperelasticity" example should also be advanced, right?

> ## List of examples (chronological order)

Author: ```suggestion
## List of examples (sequential order)
```
My intention was to create a list with the same order as on the sidebar of the homepage. I changed the desciption to match that, though it may be possible examples later in the list were added earlier (in time).

	
 Keywords: Topology optimization, weak and strong form, non-linear problem, Laplacian, grid topology

## Other examples
1.	[Incompressible Navier-Stokes Equations via DifferentialEquations.jl](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/ns_vs_diffeq/)  

 The weak form of the semi-discretized incompressible Navier-Stokes equations are derived from the strong form and implemented in the finite element code. Then, the time-dependent solution is calculated by using the DifferentialEquations.jl package. For this purpose, the PDE is required in a specific form.  
	
 Keywords: Fluid dynamics, weak and strong form, solver, incompressibility, time-dependent problem

## Grouped by topic
* Fundamentals: [Heat Equation](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/heat_equation/), [Postprocessing](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/postprocessing/), [Helmholtz Equation](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/helmholtz/), [Incompressible Elasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/incompressible_elasticity/)

> * Non-linear Problems: [Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/hyperelasticity/), [von Mises Plasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/plasticity/), [Nearly Incompressible Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/quasi_incompressible_hyperelasticity/), [Topology Optimization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/topology_optimization/)

Reviewer: Computational homogenization is missing.

Author: ```suggestion
* Non-linear Problems: [Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/hyperelasticity/), [von Mises Plasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/plasticity/), [Nearly Incompressible Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/quasi_incompressible_hyperelasticity/), [Computational homogenization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/computational_homogenization/), [Topology Optimization](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/topology_optimization/)
```

> * Time dependent problems: [Time Dependent Problems](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/transient_heat_equation/), [Incompressible Navier-Stokes Equations via DifferentialEquations.jl](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/ns_vs_diffeq/)

Reviewer: Actually the first problem should be "transient heat equation" right? @koehlerson

Author: I would agree that this would be more discriptive, I just used the titles used in the sidebar to match and avoid confusion.

* Multiple fields: [Incompressible Elasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/incompressible_elasticity/), [Linear shell](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/linear_shell/), [Nearly Incompressible Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/quasi_incompressible_hyperelasticity/)

> * Coupled problems: [Stokes flow](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/stokes-flow/)     

Reviewer: Stokes flow is not a coupled problem, but a "multiple field" problem.

Author: Then we currently don't have a "true" coupled problem and the coupled problem list is obsolete for now.

> * Multiple fields: [Incompressible Elasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/incompressible_elasticity/), [Linear shell](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/linear_shell/), [Nearly Incompressible Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/quasi_incompressible_hyperelasticity/)
> * Coupled problems: [Stokes flow](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/stokes-flow/)     

Author: ```suggestion
* Multiple fields: [Incompressible Elasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/incompressible_elasticity/), [Linear shell](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/linear_shell/), [Nearly Incompressible Hyperelasticity](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/quasi_incompressible_hyperelasticity/), [Stokes flow](https://ferrite-fem.github.io/Ferrite.jl/dev/examples/stokes-flow/)     
```
